### PR TITLE
[mlir][vector] Implement speculation for vector.transferx ops

### PR DESCRIPTION
**Pull request by Groverkss**
This patch implements speculation for vector.transfer_read/vector.transfer_write ops, allowing these ops to work with LICM.

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
What happens with out-of-bounds xfers over tensors?

**Comment by Groverkss** [comment]
> What happens with out-of-bounds xfers over tensors?

I think it's still speculatable. The op has clearly defined out of bounds behavior, which does not cause any side effects.

**Review by kuhar — APPROVED** [review]
LGTM. Should we also cover gather in the same sweep?

**Comment by Groverkss** [comment]
> LGTM. Should we also cover gather in the same sweep?

I thought about doing gather/load/store/scatter also, but it looks like they don't have MemoryEffectsOpInterface implemented also. I'm going to defer it for now, since I need to see how to test that interface.

**Comment by kuhar** [comment]
> I thought about doing gather/load/store/scatter also, but it looks like they don't have MemoryEffectsOpInterface implemented also. I'm going to defer it for now, since I need to see how to test that interface.

I think `vector.gather` is the only other op that accepts tensors, no?

**Comment by banach-space** [comment]
Not really an expert, but going over https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/, it's not clear to me why these Ops fall under `ConditionallySpeculatable`? Could you expand a bit? 

I just want to make sure that this is not working around limitations in the LICM logic (which I am not familiar with).

**Comment by Groverkss** [comment]
> Not really an expert, but going over https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/, it's not clear to me why these Ops fall under `ConditionallySpeculatable`? Could you expand a bit?
> 
> I just want to make sure that this is not working around limitations in the LICM logic (which I am not familiar with).

I can give an overview of my reasoning.

LICM can only hoist operations (atleast for now) that do not have any memory side effects and are speculatable (do not cause undefined behavior).

Operations working on tensors (linalg.generic, vector.transfer_*, etc.) operations have no memory side effects and don't have undefined behavior generally (atleast from my understanding of tensors having value semantics). So when operating on tensors, these operations are "Pure".

Operations working on memrefs (memref versions of linalg.generic, vector.transfer_*, etc.) operations have clearly defined memory side effects but may introduce out of bound behavior. So they are non-speculatable (atleast without any further analysis).

Since linalg.generic/vector.transfer_... ops work on both memref and tensors, they have to be "ConditionallySpeculatable", the condition being if they have memref or tensor semantics.

I don't think this is working around limitations of LICM, rather it is how LICM expects operations to give it information to work.

**Comment by Groverkss** [comment]
> > I thought about doing gather/load/store/scatter also, but it looks like they don't have MemoryEffectsOpInterface implemented also. I'm going to defer it for now, since I need to see how to test that interface.
> 
> I think `vector.gather` is the only other op that accepts tensors, no?

interesting... I can send a follow up pr for that.

**Comment by banach-space** [comment]
Thanks for elaborating, that helps!

> Because executing them speculatively is guaranteed not to have any 'implicit behavior' (side effects / trigger UB / etc.): https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/#categorization.


A lot (most?) things in that doc only really matter when dealing with MemRefs. And the implementation here also implies that:
* when operating on tensors, everything is safe,
* when operating on memrefs, exercise caution.

This just makes me wonder, shouldn't LICM be checking for tensor semantics instead?

> Since linalg.generic/vector.transfer_... ops work on both memref and tensors, they have to be "ConditionallySpeculatable", the condition being if they have memref or tensor semantics.

Or,  we are missing sth like `SpeculableWhenUsingTensorSemantics`.

File: mlir/test/Transforms/loop-invariant-code-motion.mlir
Comment (banach-space) [inline mlir/test/Transforms/loop-invariant-code-motion.mlir:1277]: [nit] The name of this func test is misleading -  no hoisting happens.

Also, could you add a negative test with tensor semantics. Just to make sure that we don't sudddenly start hoisting stuff only because it operates on tensors.

**Comment by Groverkss** [comment]
> Thanks for elaborating, that helps!
> 
> > Because executing them speculatively is guaranteed not to have any 'implicit behavior' (side effects / trigger UB / etc.): https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/#categorization.
> 
> A lot (most?) things in that doc only really matter when dealing with MemRefs. And the implementation here also implies that:
> 
> * when operating on tensors, everything is safe,
> * when operating on memrefs, exercise caution.
> 
> This just makes me wonder, shouldn't LICM be checking for tensor semantics instead?
> 
> > Since linalg.generic/vector.transfer_... ops work on both memref and tensors, they have to be "ConditionallySpeculatable", the condition being if they have memref or tensor semantics.
> 
> Or, we are missing sth like `SpeculableWhenUsingTensorSemantics`.

I think "SpeculableWhenUsingTensorSemantics" is exactly what's missing. Although i'm not sure if it's that simple because we would also need some way of saying when the op has tensor semantics.

**Comment by kuhar** [comment]
> A lot (most?) things in that doc only really matter when dealing with MemRefs. And the implementation here also implies that:
> when operating on tensors, everything is safe,
> when operating on memrefs, exercise caution.

This is not true in the general case outside of these ops. For example, dividing a tensor by tensor (elementwise) is not always safe. I think what we have with these 'element access ops' here might be a special case.

**Comment by banach-space** [comment]
@Groverkss You have landed this change without all of the active reviewers explicitly approving it, please avoid that. As per [contribution guidelines](https://llvm.org/docs/Contributing.html#how-to-submit-a-patch):

> Please address the feedback and re-post an updated version of your patch. This cycle continues until all requests and comments have been addressed and a reviewer accepts the patch with a Looks good to me or LGTM. Once that is done the change can be committed. 

I read that as "if a reviewer leaves a comment, please address that and wait for them to confirm that their comments have been addressed by posting **LGTM**". That last part didn't happen, hence this message.

I'm not an expert on this area, but this changes looks good to me and I appreciate you did address my comments. Still, I'd appreciate if you followed the process in the future, thanks!

**Comment by kuhar** [comment]
@banach-space overall, I agree with you as what the best practice should be, but my reading of the guidelines is different:

> This cycle continues until all requests and comments have been addressed and a reviewer accepts the patch with a Looks good to me or LGTM

This presents two criteria: (1) there are no outstanding comments that require resolution, (2) the PR received at least on LGTM. It doesn't require you to receive a LGTM from all reviewers / commenters. IME, very often this is a judgement call of whether you reasonably believe that all comments have been addressed, and I see waiting for a confirmation is more of a good practice.

**Comment by Groverkss** [comment]
> @Groverkss You have landed this change without all of the active reviewers explicitly approving it, please avoid that. As per [contribution guidelines](https://llvm.org/docs/Contributing.html#how-to-submit-a-patch):
> 
> > Please address the feedback and re-post an updated version of your patch. This cycle continues until all requests and comments have been addressed and a reviewer accepts the patch with a Looks good to me or LGTM. Once that is done the change can be committed.
> 
> I read that as "if a reviewer leaves a comment, please address that and wait for them to confirm that their comments have been addressed by posting **LGTM**". That last part didn't happen, hence this message.
> 
> I'm not an expert on this area, but this changes looks good to me and I appreciate you did address my comments. Still, I'd appreciate if you followed the process in the future, thanks!

Ah sorry, I misunderstood the only review comment being a minor comment about tests as a signal that it should be fine to land once it's fixed as you seemed satisfied with the explanation provided before. But I should have taken more caution and waited, as I wasn't in any hurry to land it and it's always nicer to get more eyes here. I really appreciate the review and hopefully, this clears why I landed it and I'm happy to still address any post review comments :)

**Comment by kuhar** [comment]
> @banach-space overall, I agree with you as what the best practice should be, but my reading of the guidelines is different:
> 
> > This cycle continues until all requests and comments have been addressed and a reviewer accepts the patch with a Looks good to me or LGTM
> 
> This presents two criteria: (1) there are no outstanding comments that require resolution, (2) the PR received at least one LGTM. It doesn't require you to receive a LGTM from all reviewers / commenters. IME, very often this is a judgement call of whether you reasonably believe that all comments have been addressed, and I see waiting for a confirmation as more of a good practice.

For example, you typically wouldn't wait for approval from all 'drive-by' reviewers who point out typos and other nits.

**Comment by kuhar** [comment]
@banach-space this is a pretty good breakdown, maybe it would be worth spelling this out in the guidelines like that?

**Pull request merged**
